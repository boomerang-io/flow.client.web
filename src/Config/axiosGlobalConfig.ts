// Setup global config for Axios requests so they work with ICP
import axios from "axios";
axios.defaults.withCredentials = true;
