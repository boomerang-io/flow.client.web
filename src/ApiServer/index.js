import { Server, Serializer, Model, Response } from "miragejs";
import { inflections } from "inflected";
import queryString from "query-string";
import { v4 as uuid } from "uuid";
import { serviceUrl, BASE_URL } from "Config/servicesConfig";
import * as fixtures from "./fixtures";

export function startApiServer({ environment = "test", timing = 0 } = {}) {
  inflections("en", function (inflect) {
    // Prevent pluralization bc our apis are weird
    inflect.irregular("activity", "activity");
    inflect.irregular("config", "config");
    // inflect.irregular("tasktemplates", "tasktemplates");
    inflect.irregular("insights", "insights");
    inflect.irregular("flowNavigation", "flowNavigation");
  });

  return new Server({
    environment,
    // Load in mock data
    fixtures,
    // Return the data as is, don't add a root key
    serializers: {
      application: Serializer.extend({
        root: false,
        embed: true,
      }),
    },
    // Register the data as a model so we can use the schema
    models: {
      activity: Model,
      approverGroups: Model,
      changelog: Model,
      globalParams: Model,
      featureFlag: Model,
      insights: Model,
      manageTeam: Model,
      manageTeamDetail: Model,
      manageUser: Model,
      quotas: Model,
      revision: Model,
      setting: Model,
      summary: Model,
      taskTemplates: Model,
      team: Model,
      teamApproverUsers: Model,
      teamNameValidate: Model,
      taskTemplateValidate: Model,
      teamProperties: Model,
      tokens: Model,
      flowNavigation: Model,
      workflowCalendar: Model,
      workflowSchedules: Model,
    },

    routes() {
      // Control how long the responses take to resolve
      this.timing = timing;

      // Allow unhandled requests on the current domain to pass through
      this.passthrough();

      this.get("/info", () => []);

      /**
       * Simple GET of static data
       */
      this.get(serviceUrl.getUserProfile(), (schema) => {
        return schema.db.profile[0];
      });

      this.get(serviceUrl.getContext(), (schema) => {
        return schema.db.platformConfig[0];
      });

      this.get(`${BASE_URL}/navigation`, (schema) => {
        return schema.db.flowNavigation;
      });

      this.get(serviceUrl.getMyTeams({ query: null }), (schema) => {
        return schema.db.myTeams[0];
      });

      this.get(serviceUrl.getFeatureFlags(), (schema) => {
        return schema.db.featureFlags[0];
      });

      this.get(
        serviceUrl.getWorkflowSchedulesCalendar({
          workflowId: ":workflowId",
          query: null,
        }),
        (schema) => {
          return schema.db.workflowCalendar;
        }
      );

      this.get(serviceUrl.getSchedule({ query: null }), (schema) => {
        return schema.db.workflowSchedules;
      });

      this.get(serviceUrl.getSchedules({ query: null }), (schema) => {
        return schema.db.workflowSchedules[0];
      });

      this.get(serviceUrl.getWorkflowSchedules({ workflowId: ":workflowId" }), (schema) => {
        return schema.db.workflowSchedules;
      });

      this.get(serviceUrl.workflowAvailableParameters({ workflowId: ":workflowId" }), (schema) => {
        return schema.db.availableParameters[0].data;
      });

      this.get(serviceUrl.getWorkflowTemplates(), (schema) => {
        return schema.db.workflowTemplates[0];
      });

      this.get(serviceUrl.getWorkflows({ query: null }), (schema) => {
        return schema.db.workflows[0];
      });

      this.get(serviceUrl.getTeamQuotas({ id: ":id" }), (schema) => {
        return schema.db.quotas[0];
      });

      this.get(serviceUrl.getTeams({ query: null }), (schema) => {
        return schema.db.teams[0];
      });

      this.put(serviceUrl.putActivationApp(), () => {
        return {};
      });

      /**
       * Global Parameters
       */

      this.get(serviceUrl.getGlobalParams({ query: null }), (schema) => {
        return schema.db.globlaParams[0];
      });
      this.post(serviceUrl.getGlobalParams(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        schema.globalParams.create({ id: uuid(), ...body });
        return schema.globalParams.all();
      });

      this.patch(serviceUrl.getGlobalProperty({ id: ":id" }), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        let { id } = request.params;
        let param = schema.globalParams.find(id);
        param.update({ ...body });
      });

      this.delete(serviceUrl.getGlobalProperty({ id: ":id" }), (schema, request) => {
        let { id } = request.params;
        schema.db.globalParams.remove({ id });
      });

      /**
       * Team Propertiies
       */
      this.get(serviceUrl.getTeamProperties({ id: ":id" }), (schema, request) => {
        let { id } = request.params;
        let property = schema.teamProperties.find(id);
        return property && property.properties ? property.properties : [];
      });
      this.post(serviceUrl.getTeamProperties({ id: ":id" }), (schema, request) => {
        /**
         * find team record, update the list of properties for that team
         */
        let { id } = request.params;
        let body = JSON.parse(request.requestBody);
        let activeTeamProperty = schema.teamProperties.find(id);
        let currentProperties = activeTeamProperty.attrs.properties;
        currentProperties.push({ id: uuid(), ...body });
        activeTeamProperty.update({ properties: currentProperties });
        return schema.teamProperties.all();
      });
      this.patch(
        serviceUrl.getTeamProperty({ teamId: ":teamId", configurationId: ":configurationId" }),
        (schema, request) => {
          /**
           * find team record, update the list of properties for that team
           */
          let { teamId, configurationId } = request.params;
          let body = JSON.parse(request.requestBody);
          let activeTeamProperty = schema.teamProperties.find(teamId);
          let currentProperties = activeTeamProperty.attrs.properties;
          let foundIndex = currentProperties.findIndex((prop) => prop.id === configurationId);
          currentProperties[foundIndex] = body;
          activeTeamProperty.update({ properties: currentProperties });
          return schema.teamProperties.all();
        }
      );
      this.delete(
        serviceUrl.getTeamProperty({ teamId: ":teamId", configurationId: ":configurationId" }),
        (schema, request) => {
          /**
           * find team record, update the list of properties for that team
           */
          let { teamId, configurationId } = request.params;
          let activeTeamProperty = schema.teamProperties.find(teamId);
          let currentProperties = activeTeamProperty.attrs.properties;
          let newProperties = currentProperties.filter((prop) => prop.id !== configurationId);
          activeTeamProperty.update({ properties: newProperties });
          return schema.teamProperties.all();
        }
      );

      /**
       * Insights
       */
      this.get(serviceUrl.getInsights({ query: null }), (schema, request) => {
        //grab the querystring from the end of the request url
        const query = request.url.substring(14);
        // eslint-disable-next-line
        const { fromDate = null, toDate = null, teamId = null } = queryString.parse(query);
        const activeTeam = teamId && schema.db.myTeams.find(teamId);
        let activeExecutions =
          activeTeam && schema.db.insights[0].executions.filter((team) => team.teamName === activeTeam.name);
        return activeExecutions ? { ...schema.db.insights[0], executions: activeExecutions } : schema.db.insights[0];
      });

      /**
       * Task Templates
       */
      const tasktemplatePath = serviceUrl.getTaskTemplates({ query: null });
      this.get(serviceUrl.getTaskTemplates({ query: null }), (schema) => {
        console.log(schema.db.taskTemplates);
        return schema.db.taskTemplates[0];
      });
      this.put(tasktemplatePath, (schema, request) => {
        let body = JSON.parse(request.requestBody);
        let taskTemplate = schema.tasktemplate.find(body.id);
        taskTemplate.revisions.push(body);
        taskTemplate.update({ ...body });
        return taskTemplate;
      });

      this.post(serviceUrl.postValidateYaml(), (schema) => {
        return schema.db.taskTemplateValidate[0];
      });

      /**
       * Workflows
       */

      // Workflow Summary
      this.get(serviceUrl.getWorkflowSummary({ workflowId: ":workflowId" }), (schema, request) => {
        let { workflowId } = request.params;
        return schema.summaries.find(workflowId);
      });

      this.patch(serviceUrl.patchUpdateWorkflowSummary(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        let summary = schema.summaries.find(body.workflowId);
        summary.update(body);
        return summary;
      });

      this.post(serviceUrl.postCreateWorkflow(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        let workflow = { ...body, id: uuid(), createdDate: Date.now(), revisionCount: 1, status: "active" };
        if (body.flowTeamId) {
          let flowTeam = schema.myTeams.findBy({ id: body.flowTeamId });
          const teamWorkflows = [...flowTeam.workflows];
          teamWorkflows.push(workflow);
          flowTeam.update({ workflows: teamWorkflows });
          return schema.summaries.create(workflow);
        }
        return {};
      });

      this.post(
        serviceUrl.postCreateWorkflowToken({ workflowId: ":workflowId", label: ":label" }),
        (schema, request) => {
          return { token: uuid() };
        }
      );

      this.del(serviceUrl.getWorkflow({ id: ":workflowId" }), (schema, request) => {
        let { workflowId } = request.params;
        let flowTeam = schema.myTeams.where((team) => team.workflows.find((workflow) => workflow.id === workflowId));
        let { attrs } = flowTeam.models[0];
        const teamWorkflows = attrs.workflows.filter((workflow) => workflow.id !== workflowId);
        flowTeam.update({ workflows: teamWorkflows });
        return schema.db.summaries.remove({ id: workflowId });
      });

      // Workflow Revision
      this.get(serviceUrl.getWorkflowRevision({ workflowId: ":workflowId" }), (schema, request) => {
        let { workflowId } = request.params;
        if (workflowId) {
          return schema.revisions.findBy({ workFlowId: workflowId });
        } else {
          return {};
        }
      });

      this.get(
        serviceUrl.getWorkflowRevision({ workflowId: ":workflowId", revisionNumber: ":revisionNumber" }),
        (schema, request) => {
          let { workflowId, revisionNumber } = request.params;
          if (revisionNumber) {
            return schema.revisions.findBy({ workFlowId: workflowId, version: revisionNumber });
          } else {
            return {};
          }
        }
      );

      this.get(serviceUrl.getWorkflowTaskTemplates({ workflowId: ":workflowId" }), (schema, request) => {
        return schema.db.tasktemplate;
      });

      this.post(serviceUrl.postCreateWorkflowRevision({ workflowId: ":workflowId" }), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        let { workflowId } = request.params;
        let revision = { ...body, workFlowId: workflowId };
        return schema.revisions.create(revision);
      });

      //Workflow Config Cron

      this.get(`${BASE_URL}/workflow/validate/cron`, () => {
        return {
          valid: true,
        };
      });

      // Workflow Properties
      this.patch(serviceUrl.patchUpdateWorkflowProperties({ workflowId: ":workflowId" }), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        let { workflowId } = request.params;
        let summary = schema.summaries.find(workflowId);
        summary.update({ properties: body });
        return summary;
      });

      // Workflow Changelog
      this.get(serviceUrl.getWorkflowChangelog({ workflowId: ":workflowId" }), (schema, request) => {
        let { workflowId } = request.params;
        return schema.changelogs.where({ workflowId });
      });

      //Workflow Available Parameters
      this.post(serviceUrl.workflowAvailableParameters({ workflowId: ":workflowId" }), (schema, request) => {
        return schema.db.availableParameters[0].data;
      });
      /**
       * Activity
       */
      this.get(serviceUrl.getWorkflowRuns({ query: null }), (schema) => {
        return schema.db.workflowRuns[0];
      });

      this.get(serviceUrl.getWorkflowRunCount({ query: null }), (schema, request) => {
        return schema.db.workflowRunCount[0];
      });

      this.get(serviceUrl.getWorkflowExecution({ executionId: ":id" }), (schema, request) => {
        return schema.db.workflowExecution[0];
      });

      this.delete(serviceUrl.deleteCancelWorkflow({ executionId: ":id" }), (schema, request) => {
        return {};
      });

      /**
       * Actions
       */
      this.get(serviceUrl.getActionsSummary({ query: null }), (schema) => {
        return schema.db.actionsSummary[0];
      });

      this.get(serviceUrl.getActions({ query: null }), (schema, request) => {
        const { type } = request.queryParams;
        if (type === "approval") return schema.db.approvals[0];
        if (type === "task") return schema.db.manualTasks[0];
        return {};
      });

      this.put(serviceUrl.putWorkflowAction(), () => {
        return {};
      });

      /**
       * Approvers Group
       */
      this.get(serviceUrl.resourceApproverGroups({ teamId: ":teamId" }), (schema) => {
        return schema.db.approverGroups;
      });

      this.get(serviceUrl.getTeamMembers({ teamId: ":teamId" }), (schema) => {
        return schema.db.teamApproverUsers;
      });

      //Delete approver group
      this.delete(serviceUrl.resourceApproverGroups({ teamId: ":teamId", groupId: ":groupId" }), (schema, request) => {
        const { groupId } = request.params;
        const approverGroup = schema.approverGroups.find(groupId);
        approverGroup.destroy();
      });

      //Create approver group
      this.post(serviceUrl.resourceApproverGroups({ teamId: ":teamId" }), (schema, request) => {
        const body = JSON.parse(request.requestBody);
        schema.approverGroups.create({ groupId: uuid(), ...body });
        return schema.approverGroups.all();
      });

      //Update approver group
      this.put(serviceUrl.resourceApproverGroups({ teamId: ":teamId" }), (schema, request) => {
        return {};
      });

      /**
       * Manage Team
       */

      this.post(serviceUrl.postTeamValidateName(), (schema, request) => {
        return new Response(422, {}, { errors: [ 'Name is already taken'] });
      });

      this.get(serviceUrl.getTeam({ teamId: ":teamId" }), (schema, request) => {
        let { teamId } = request.params;
        return schema.manageTeamDetails.findBy({ id: teamId });
      });

      this.patch(serviceUrl.getTeam({ teamId: ":teamId" }), (schema, request) => {
        let { teamId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activeTeam = schema.manageTeamDetails.findBy({ id: teamId });
        let activeUsers = activeTeam.users.filter((user) => body.includes(user.id));
        activeTeam.update({ users: activeUsers });
        return activeTeam;
      });

      this.put(serviceUrl.getTeam({ teamId: ":teamId" }), (schema, request) => {
        let { teamId } = request.params;
        let body = JSON.parse(request.requestBody);
        let summary = schema.manageTeamDetails.findBy({ id: teamId });
        summary.update(body);
        return summary;
      });

      this.patch(serviceUrl.getManageTeamLabels({ teamId: ":teamId" }), (schema, request) => {
        let { teamId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activeTeam = schema.manageTeamDetails.findBy({ id: teamId });
        activeTeam.update({ labels: body });
        return activeTeam;
      });

      this.post(serviceUrl.getManageTeamsCreate(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        const teams = schema.teams.first();
        const updatedRecords = teams.records.concat({ id: uuid(), isActive: true, ...body });
        teams.update({ records: updatedRecords });
        return {};
      });

      /**
       * Manage Users
       */

      this.get(serviceUrl.getUsers({ query: null }), (schema, request) => {
        const { query } = request.queryParams;
        const userData = schema.db.users[0];
        if (query) {
          userData.content =
            userData.content.filter((user) => user.name.includes(query) || user.email.includes(query)) ?? [];
        }
        return userData;
      });

      this.get(serviceUrl.getUser({ userId: ":userId" }), (schema, request) => {
        const { userId } = request.params;
        const user = schema.db.users[0].content.find((user) => user.id === userId);
        return user;
      });

      this.patch(serviceUrl.getUser({ userId: ":userId" }), (schema, request) => {
        const { userId } = request.params;
        let body = JSON.parse(request.requestBody);
        const users = schema.users.first();
        const updatedRecords = [];
        for (let user of users.records) {
          if (user.id === userId) {
            user = user = { ...user, ...body };
          }
          updatedRecords.push(user);
        }
        users.update({ records: updatedRecords });
        return {};
      });

      /**
       * Quotas
       */

      this.patch(serviceUrl.getTeamQuotas({ id: ":id" }), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        const quotas = schema.quotas.first();

        const key = Object.keys(body)[0];
        const value = Object.values(body)[0];

        quotas.update({ [key]: value });
        return {};
      });

      this.put(serviceUrl.putTeamQuotasDefault({ id: ":id" }), (schema, request) => {
        const quotas = schema.quotas.first();
        const defaultConfig = {
          maxWorkflowCount: 20,
          maxWorkflowExecutionMonthly: 150,
          maxWorkflowStorage: 10,
          maxWorkflowExecutionTime: 30,
          maxConcurrentWorkflows: 4,
        };

        quotas.update(defaultConfig);
        return {};
      });

      this.put(serviceUrl.getTeamQuotas({ id: ":id" }), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        const quotas = schema.quotas.first();
        quotas.update(body);
        return {};
      });

      this.get(serviceUrl.getTeamQuotaDefaults(), (schema, request) => {
        return {
          maxWorkflowCount: 20,
          maxWorkflowExecutionMonthly: 150,
          maxWorkflowStorage: 10,
          maxWorkflowExecutionTime: 30,
          maxConcurrentWorkflows: 4,
        };
      });

      /**
       *  Manage Settings
       * */

      this.get(serviceUrl.resourceSettings(), (schema) => {
        return schema.settings.all();
      });

      this.put(serviceUrl.resourceSettings(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        const settings = schema.settings.all();
        settings.update(body[0]);
        return schema.settings.all();
      });

      /**
       * Manage and Administer Tokens
       */
      this.get(serviceUrl.getTeamTokens({ teamId: ":teamId" }), (schema) => {
        return schema.db.tokens;
      });

      this.get(serviceUrl.getGlobalTokens(), (schema) => {
        return schema.db.tokens;
      });

      this.delete(serviceUrl.deleteToken({ tokenId: ":tokenId" }), (schema, request) => {
        return {};
      });

      this.post(serviceUrl.postGlobalToken(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        let newToken = {
          ...body,
          creatorId: "1",
          creationDate: Date.now(),
          creatorName: "Test User",
          tokenValue: "testglobal",
        };
        return schema.tokens.create(newToken);
      });

      this.post(serviceUrl.postTeamToken(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        let newToken = {
          ...body,
          creatorId: "1",
          creationDate: Date.now(),
          creatorName: "Test User",
          tokenValue: "testteam",
        };
        return schema.tokens.create(newToken);
      });

      /**
       * Workflow Templates
       */
      this.post(serviceUrl.postDuplicateWorkflow({ workflowId: ":workflowId" }), (schema, request) => {
        return {};
      });
      /**
       * TODO
       */
    },
  });
}
