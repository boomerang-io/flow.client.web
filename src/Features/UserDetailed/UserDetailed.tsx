import React from "react";
import { Helmet } from "react-helmet";
import { useQuery } from "react-query";
import { useFeature } from "flagged";
import { Switch, Route, useRouteMatch } from "react-router-dom";
import { Box } from "reflexbox";
import { ErrorMessage, Loading } from "@carbon/react";
import Header from "./Header";
import Labels from "./Labels";
import Teams from "./Teams";
import Workflows from "./Workflows";
import { FlowUser } from "Types";
import { AppPath, FeatureFlag } from "Config/appConfig";
import { serviceUrl, resolver } from "Config/servicesConfig";
import styles from "./UserDetailed.module.scss";

interface FeatureLayoutProps {
  isError?: boolean;
  isLoading?: boolean;
  user?: FlowUser;
  children: any;
}

const FeatureLayout = ({ children, isLoading, isError }: FeatureLayoutProps) => {
  return (
    <>
      <Helmet>
        <title>User</title>
      </Helmet>
      <Header isError={isError} isLoading={isLoading} />
      <Box p="1rem">{children}</Box>
    </>
  );
};

function TeamDetailedContainer() {
  const userManagementEnabled = useFeature(FeatureFlag.UserManagementEnabled);
  const match: { params: { userId: string } } = useRouteMatch();
  const userId = match?.params?.userId;

  const userDetailsUrl = serviceUrl.resourceManageUser({ userId });

  const { data: userDetailsData, isError: userDetailsIsError, isLoading: userDetailsIsLoading } = useQuery({
    queryKey: userDetailsUrl,
    queryFn: resolver.query(userDetailsUrl),
  });

  if (userDetailsIsLoading)
    return (
      <FeatureLayout isLoading={userDetailsIsLoading}>
        <Loading />
      </FeatureLayout>
    );
  if (userDetailsIsError)
    return (
      <FeatureLayout isError={userDetailsIsError}>
        <ErrorMessage />
      </FeatureLayout>
    );

  if (userDetailsData) {
    return (
      <div className={styles.container}>
        <Header user={userDetailsData} userManagementEnabled={userManagementEnabled} />
        <Switch>
          <Route exact path={AppPath.User}>
            <Workflows user={userDetailsData} />
          </Route>
          <Route exact path={AppPath.UserTeams}>
            <Teams user={userDetailsData} />
          </Route>
          <Route exact path={AppPath.UserLabels}>
            <Labels user={userDetailsData} userManagementEnabled={userManagementEnabled} />
          </Route>
        </Switch>
      </div>
    );
  }

  return null;
}

export default TeamDetailedContainer;
