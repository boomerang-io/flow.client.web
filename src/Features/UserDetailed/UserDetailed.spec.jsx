import UserDetailed from ".";
import { Route } from "react-router-dom";
import { waitFor, screen } from "@testing-library/react";
import { AppPath, appLink } from "Config/appConfig";
import { startApiServer } from "ApiServer";

let server;

beforeEach(() => {
  server = startApiServer();
});

afterEach(() => {
  server.shutdown();
});

describe("UserDetailed --- Snapshot Test", () => {
  it("Capturing Snapshot of UserDetailed", async () => {
    const { baseElement } = global.rtlContextRouterRender(
      <Route path={AppPath.User}>
        <UserDetailed />
      </Route>,
      { route: appLink.user({ userId: "5f170b3df6ab327e302cb0a5" }) }
    );
    await screen.findByText("These are Tim Bula's workflows");
    expect(baseElement).toMatchSnapshot();
    await waitFor(() => null);
  });
});
