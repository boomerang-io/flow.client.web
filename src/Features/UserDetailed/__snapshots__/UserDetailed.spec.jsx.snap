// Vitest Snapshot v1

exports[`UserDetailed --- Snapshot Test > Capturing Snapshot of UserDetailed 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="_container_86126e"
    >
      <header
        class="bx--bmrg-feature-header _container_b9e313 --bordered"
      >
        <section
          class="bx--bmrg-feature-header__content"
        >
          <div
            class="bx--bmrg-feature-header__nav"
          >
            <nav
              aria-label="Breadcrumb"
            >
              <ol
                class="bx--breadcrumb bx--breadcrumb--no-trailing-slash"
              >
                <li
                  class="bx--breadcrumb-item"
                >
                  <a
                    class="bx--link"
                    href="/admin/users"
                  >
                    Users
                  </a>
                </li>
                <li
                  class="bx--breadcrumb-item bx--breadcrumb-item--current"
                >
                  <p
                    class="bx--link"
                  >
                    Tim Bula
                  </p>
                </li>
              </ol>
            </nav>
          </div>
          <div
            class="_infoContainer_b9e313"
          >
            <div
              class="_userContainer_b9e313"
            >
              <img
                alt="Avatar for undefined"
                class="_userAvatar_b9e313 --medium"
                src="/api/users/image/trbula@us.ibm.com"
                user="trbula@us.ibm.com"
              />
              <h1
                class="bx--bmrg-feature-header-title"
                style="margin: 0px 1rem 0px 0px;"
                title="Tim Bula"
              >
                Tim Bula
              </h1>
              <div
                class="_headerSubtitle_b9e313"
                title="trbula@us.ibm.com"
              >
                <div
                  class="_status_b9e313"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    style="fill: #009d9a;"
                    viewBox="0 0 32 32"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M13 24L4 15 5.414 13.586 13 21.171 26.586 7.586 28 9 13 24z"
                    />
                  </svg>
                  <p
                    class="_statusText_b9e313"
                  >
                    Active
                  </p>
                </div>
                <span
                  class="_statusDivider_b9e313"
                >
                  -
                </span>
                trbula@us.ibm.com
              </div>
            </div>
            <div
              class="_userDetailsContainer_b9e313"
            >
              <section
                class="_subHeaderContainer_b9e313"
              >
                <dl
                  class="_detailedInfoContainer_b9e313"
                >
                  <dt
                    class="_dataTitle_b9e313"
                  >
                    Platform Role
                  </dt>
                  <dd
                    class="_dataValue_b9e313"
                  >
                    Admin
                  </dd>
                </dl>
                <dl
                  class="_detailedInfoContainer_b9e313"
                >
                  <dt
                    class="_dataTitle_b9e313"
                  >
                    Date Joined
                  </dt>
                  <dd
                    class="_dataValue_b9e313"
                  >
                    2020-07-21 3:35pm
                  </dd>
                </dl>
                <dl
                  class="_detailedInfoContainer_b9e313"
                >
                  <dt
                    class="_dataTitle_b9e313"
                  >
                    Last Login
                  </dt>
                  <dd
                    class="_dataValue_b9e313"
                  >
                    2020-07-21 3:35pm
                  </dd>
                </dl>
              </section>
              <section
                class="_actionButtonsContainer_b9e313"
              >
                <button
                  class="bx--btn bx--btn--md bx--btn--ghost"
                  tabindex="0"
                  type="button"
                >
                  Change role
                  <svg
                    aria-hidden="true"
                    class="bx--btn__icon"
                    fill="currentColor"
                    focusable="false"
                    height="24"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M16 4a5 5 0 11-5 5 5 5 0 015-5m0-2a7 7 0 107 7A7 7 0 0016 2zM26 30H24V25a5 5 0 00-5-5H13a5 5 0 00-5 5v5H6V25a7 7 0 017-7h6a7 7 0 017 7z"
                    />
                  </svg>
                </button>
              </section>
            </div>
          </div>
          <div
            class="bx--bmrg-feature-header__footer"
          >
            <nav
              class="bx--bmrg-feature-nav-tabs"
            >
              <a
                aria-current="page"
                class="bx--bmrg-feature-nav-tab bx--bmrg-feature-nav-tab--active"
                href="/admin/users/5f170b3df6ab327e302cb0a5"
              >
                Workflows
              </a>
              <a
                class="bx--bmrg-feature-nav-tab"
                href="/admin/users/5f170b3df6ab327e302cb0a5/teams"
              >
                Teams
              </a>
              <a
                class="bx--bmrg-feature-nav-tab"
                href="/admin/users/5f170b3df6ab327e302cb0a5/labels"
              >
                Labels
              </a>
            </nav>
          </div>
        </section>
      </header>
      <section
        aria-label="Tim Bula Workflows"
        class="_container_cb85e4"
      >
        <section
          class="_actionsContainer_cb85e4"
        >
          <div
            class="_leftActions_cb85e4"
          >
            <p
              class="_featureDescription_cb85e4"
            >
              These are Tim Bula's workflows
            </p>
            <p
              class="_workflowCountText_cb85e4"
            >
              Showing 
              2
               workflow
              s
            </p>
            <div
              aria-labelledby="workflow-search-search"
              class="bx--search bx--search--xl"
              role="search"
            >
              <div
                class="bx--search-magnifier"
              >
                <svg
                  aria-hidden="true"
                  class="bx--search-magnifier-icon"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15,14.3L10.7,10c1.9-2.3,1.6-5.8-0.7-7.7S4.2,0.7,2.3,3S0.7,8.8,3,10.7c2,1.7,5,1.7,7,0l4.3,4.3L15,14.3z M2,6.5	C2,4,4,2,6.5,2S11,4,11,6.5S9,11,6.5,11S2,9,2,6.5z"
                  />
                </svg>
              </div>
              <label
                class="bx--label"
                for="workflow-search"
                id="workflow-search-search"
              >
                workflow search
              </label>
              <input
                autocomplete="off"
                class="bx--search-input"
                id="workflow-search"
                placeholder="Search for a workflow"
                role="searchbox"
                type="text"
              />
              <button
                aria-label="Clear search input"
                class="bx--search-close bx--search-close--hidden"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 32 32"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </section>
        <div
          aria-label="Structured list section"
          class="bx--structured-list"
          role="table"
        >
          <div
            class="bx--structured-list-thead"
            role="rowgroup"
          >
            <div
              class="bx--structured-list-row bx--structured-list-row--header-row"
              role="row"
            >
              <div
                class="bx--structured-list-th"
                role="columnheader"
              >
                Name
              </div>
              <div
                class="bx--structured-list-th"
                role="columnheader"
              >
                Summary
              </div>
              <div
                class="bx--structured-list-th"
                role="columnheader"
              >
                Revision Count
              </div>
              <div
                class="bx--structured-list-th"
                role="columnheader"
              />
              <div
                class="bx--structured-list-th"
                role="columnheader"
              />
            </div>
          </div>
          <div
            class="bx--structured-list-tbody"
            role="rowgroup"
          >
            <div
              class="bx--structured-list-row"
              role="row"
            >
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                <div
                  class="_workflowNameContainer_cb85e4"
                >
                  <p>
                    Personal - Java - Build
                  </p>
                </div>
              </div>
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                Java - Build
              </div>
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                16
              </div>
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                <a
                  class="_viewWorkflowLink_cb85e4"
                  href="/editor/5e877d1f4bbc6e0001c51e12/workflow"
                >
                  View/edit
                </a>
              </div>
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                <a
                  class="_viewWorkflowLink_cb85e4"
                  href="/activity?page=0&size=10&workflowIds=5e877d1f4bbc6e0001c51e12"
                >
                  Activity
                </a>
              </div>
            </div>
            <div
              class="bx--structured-list-row"
              role="row"
            >
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                <div
                  class="_workflowNameContainer_cb85e4"
                >
                  <p>
                    Personal - ML Train – Bot Efficiency
                  </p>
                </div>
              </div>
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                Train and store ML model for Bot Efficiency.
              </div>
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                2
              </div>
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                <a
                  class="_viewWorkflowLink_cb85e4"
                  href="/editor/5eb2c4085a92d80001a16d87/workflow"
                >
                  View/edit
                </a>
              </div>
              <div
                class="bx--structured-list-td"
                role="cell"
              >
                <a
                  class="_viewWorkflowLink_cb85e4"
                  href="/activity?page=0&size=10&workflowIds=5eb2c4085a92d80001a16d87"
                >
                  Activity
                </a>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>
  <div
    class="bx--bmrg-modal-portal"
  />
</body>
`;
